int-set/src/bitset.rs

```diff
@@ -184,30 +235,35 @@ impl<T> BitSet<T> {
     fn page_for(&self, value: u32) -> Option<&BitPage> {
         let major_value = self.get_major_value(value);
         let pages_index = self.page_index_for_major(major_value)?;
-        Some(&self.pages[pages_index])
+        self.pages.get(pages_index)
     }
 
     /// Return a mutable reference to the page that 'value' resides in.
     ///
     /// Insert a new page if it doesn't exist.
     fn page_for_mut(&mut self, value: u32) -> Option<&mut BitPage> {
         let major_value = self.get_major_value(value);
-        let pages_index = self.page_index_for_major(major_value)?;
-        Some(&mut self.pages[pages_index])
+        return self.page_for_major_mut(major_value);
+    }
+
+    // Return a mutable reference to the page with major value equal to major_value.
+    fn page_for_major_mut(&mut self, major_value: u32) -> Option<&mut BitPage> {
+        let page_index = self.page_index_for_major(major_value)?;
+        self.pages.get_mut(page_index)
     }
 
     /// Return a mutable reference to the page that 'value' resides in.
     ///
     /// Insert a new page if it doesn't exist.
-    fn ensure_page_for_mut(&mut self, value: u32) -> &mut BitPage {
+    fn ensure_page_for_mut(&mut self, value: u32) -> Option<&mut BitPage> {
         self.ensure_page_for_major_mut(self.get_major_value(value))
     }
 
     // Return a mutable reference to the page with major value equal to major_value.
     // Inserts a new page if it doesn't exist.
-    fn ensure_page_for_major_mut(&mut self, major_value: u32) -> &mut BitPage {
+    fn ensure_page_for_major_mut(&mut self, major_value: u32) -> Option<&mut BitPage> {
         let page_index = self.ensure_page_index_for_major(major_value);
-        &mut self.pages[page_index]
+        self.pages.get_mut(page_index)
```

Comment:
I have led you astray in this case: since it is an invariant that the page is always present, we should just be returning `&mut BitPage` and we should handle the unwrapping instead of requiring the caller to do it.

Reply:
no problem, I'll change it back. I prefer the panic because it will provide a clear signal that the invariants were violated instead of being silently ignored by the optional.